core/trie_verify.go

```diff
@@ -0,0 +1,345 @@
+	pruneHeightDiff = 15
+	pruneInterval   = 5 * time.Second
+	resendInterval  = 2 * time.Second
+	// tryAllPeersTime is the time that a block has not been verified and then try all the valid verify peers.
+	tryAllPeersTime = 15 * time.Second
+)
+
+type verifyManager struct {
+	bc             *BlockChain
+	tasks          map[common.Hash]*verifyTask
+	peers          verifyPeers
+	verifiedCache  *lru.Cache
+	allowUntrusted bool
+
+	newTaskCh chan *types.Header
+	verifyCh  chan common.Hash
+	messageCh chan verifyMessage
+}
+
+func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowUntrusted bool) *verifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &verifyManager{
+		bc:             blockchain,
+		tasks:          make(map[common.Hash]*verifyTask),
+		peers:          peers,
+		verifiedCache:  verifiedCache,
+		allowUntrusted: allowUntrusted,
+
+		newTaskCh: make(chan *types.Header),
+		verifyCh:  make(chan common.Hash),
+		messageCh: make(chan verifyMessage),
+	}
+	return vm
+}
+
+func (vm *verifyManager) mainLoop() {
+	// read disk store to initial verified cache
+	// load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	// Start verify task from H to H-11 if need.
+	vm.NewBlockVerifyTask(header)
+	pruneTicker := time.NewTicker(pruneInterval)
+	defer pruneTicker.Stop()
+	for {
+		select {
+		case h := <-vm.newTaskCh:
+			vm.NewBlockVerifyTask(h)
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			if task, ok := vm.tasks[hash]; ok {
+				delete(vm.tasks, hash)
+				close(task.terminalCh)
+			}
+		case <-pruneTicker.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Cmp(task.blockHeader.Number) == 1 &&
+					vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > pruneHeightDiff {
+					delete(vm.tasks, hash)
+					close(task.terminalCh)
+				}
+			}
+		case message := <-vm.messageCh:
+			if vt, ok := vm.tasks[message.verifyResult.BlockHash]; ok {
+				vt.messageCh <- message
+			}
+		case <-vm.bc.quit:
+			for _, task := range vm.tasks {
+				close(task.terminalCh)
+			}
+			return
+		}
+	}
+}
+
+func (vm *verifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; header != nil && i <= maxForkHeight; i++ {
+		func(hash common.Hash) {
+			// if verified cache record that this block has been verified, skip.
+			if _, ok := vm.verifiedCache.Get(hash); ok {
+				return
+			}
+			// if there already has a verify task for this block, skip.
+			if _, ok := vm.tasks[hash]; ok {
+				return
+			}
+			diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+			// if this block has no diff, there is no need to verify it.
+			var err error
+			if diffLayer == nil {
+				if diffLayer, err = vm.bc.GenerateDiffLayer(hash); err != nil {
+					log.Error("failed to get diff layer", "block", hash, "number", header.Number, "error", err)
+					return
+				}
```

Comment:
There is a case if the block is an empty block, the error is nil and difflayer is nil too. So, add another branch to test if difflayer == nil, if so, return directly.